mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp

```diff
@@ -0,0 +1,468 @@
+    sourceOffset = dBuilder.create<AddIOp>(
+        loc, sourceOffset,
+        dBuilder.create<MulIOp>(loc, inputSpanLenConstantOp, d_i));
+    sourceOffset = kBuilder.create<AddIOp>(
+        loc, sourceOffset,
+        kBuilder.create<MulIOp>(
+            loc,
+            kBuilder.create<AddIOp>(
+                loc,
+                kBuilder.create<MulIOp>(loc, k_i,
+                                        inputSpansPerMfmaInConstantOp),
+                blk_id),
+            dPerBlock));
+  }
+  return sourceOffset;
+}
+
+// **************************
+// Wmma accelerator interface
+// **************************
+
+WmmaEmitter::WmmaEmitter(WmmaInsn wmmaInsn, StringRef arch,
+                         XdlopsGemmParamsAttr tuningParams)
+    : AccelEmitter{arch, tuningParams,
+                   initAccelEmitterParams(wmmaInsn, tuningParams)},
+      wmmaInsn(wmmaInsn) {}
+
+AccelEmitterParams
+WmmaEmitter::initAccelEmitterParams(WmmaInsn wmmaInsn,
+                                    XdlopsGemmParamsAttr tuningParams) {
+  AccelEmitterParams params;
+
+  // Extract relevant tuning parameters
+  int64_t kPerBlock = tuningParams.getKPerBlock();
+  int64_t kPack = tuningParams.getKpack();
+
+  params.mRepeats = wmmaInsn.mRepeats;
+  params.nRepeats = wmmaInsn.nRepeats;
+  params.nResultVectors = 1;
+  params.kPerThread = kPerBlock;
+  params.kBase = wmmaInsn.inputLen;
+  params.mPerAccel = wmmaInsn.inputLen;
+  params.nPerAccel = wmmaInsn.inputLen;
+  params.kBasePerThread = kPerBlock * kPack / params.kBase;
+
+  params.argTypeA = wmmaInsn.argTypeA;
+  params.argTypeB = wmmaInsn.argTypeB;
+  params.accVectorType = wmmaInsn.retType;
+
+  return params;
+}
+
+Value WmmaEmitter::computeLdsSourceOffset(OpBuilder &kBuilder, Value k_i,
+                                          OpBuilder &dBuilder, Value d_i,
+                                          OpBuilder &builder, Value dPerBlock,
+                                          Location loc, Value sourceBase,
+                                          Value laneId) {
+
+  // srcOffset = k_i * MN + (laneId % wmmaInputLen) + wmmaInputLen * mn_i;
+  Value inputLen = builder.create<ConstantIndexOp>(loc, wmmaInsn.inputLen);
+  Value sourceOffset = builder.create<AddIOp>(
+      loc, sourceBase, builder.create<RemUIOp>(loc, laneId, inputLen));
+  sourceOffset = dBuilder.create<AddIOp>(
+      loc, sourceOffset, dBuilder.create<MulIOp>(loc, inputLen, d_i));
+  sourceOffset = kBuilder.create<AddIOp>(
+      loc, sourceOffset, kBuilder.create<MulIOp>(loc, dPerBlock, k_i));
+  return sourceOffset;
+}
+
+void WmmaEmitter::emitThreadwiseLoop(OpBuilder &b, Location loc, Value argA,
+                                     Value argB, Value bufferC,
+                                     Value regCOffset) {
+  VectorType vectorType = wmmaInsn.retType;
+  auto vectorC = b.create<memref::LoadOp>(loc, vectorType, bufferC, regCOffset);
+
+  auto mfma = b.create<amdgpu::WMMAOp>(loc, vectorType, argA, argB, vectorC);
```

Comment:
Yes, so the intrinsic works exactly in this way: for `fp16` it only fills half of the VGPRs they use. It can fill the top half or the bottom half. So the output is 8 VGPRs which means vector of 16 fp16 elements. If `subwordOffset` is 0, then the top half is used (hence 0, 2, .. 14) , while if it is 1, the bottom half is used.